File: src/width.ts
/**
 * Unicode display width: how many terminal columns a character occupies (0, 1, or 2).
 * Width tables derived from Unicode 17.0 data (see scripts/generate-emoji-widths.ts).
 */

import { EMOJI_PRESENTATION_RANGES, TEXT_PRESENTATION_EMOJI_RANGES } from './emoji-data.gen.js';

/** Binary search through sorted [start, end] inclusive ranges. */
function inRanges(cp: number, ranges: [number, number][]): boolean {
  let lo = 0, hi = ranges.length - 1;
  while (lo <= hi) {
    const mid = (lo + hi) >> 1;
    const [start, end] = ranges[mid]!;
    if (cp < start) hi = mid - 1;
    else if (cp > end) lo = mid + 1;
    else return true;
  }
  return false;
}

/**
 * Return the display width of a single Unicode code point.
 * Returns 2 for fullwidth/wide characters (CJK, emoji, etc).
 * Returns 0 for zero-width characters (combining marks, ZWJ, etc).
 * Returns 1 for everything else.
 */
export function charDisplayWidth(codePoint: number): number {
  // Zero-width characters
  if (isZeroWidth(codePoint)) return 0;
  // Terminal-specific overrides: these Japanese button emoji (🈂 🈷) are rendered as
  // width 2 by most terminals despite having Emoji_Presentation=No in Unicode data.
  // Verified against iTerm2, Terminal.app, and kitty via test-terminal-widths.ts.
  if (codePoint === 0x1f202 || codePoint === 0x1f237) return 2;
  // Wide / fullwidth characters
  if (isWide(codePoint)) return 2;
  return 1;
}

/**
 * Returns true if the code point is a text-presentation emoji
 * (Emoji=Yes but Emoji_Presentation=No). These are width 1 by default
 * but become width 2 when followed by U+FE0F (VS16).
 */
export function isTextPresentationEmoji(codePoint: number): boolean {
  return inRanges(codePoint, TEXT_PRESENTATION_EMOJI_RANGES);
}

function isZeroWidth(cp: number): boolean {
  // Combining Diacritical Marks
  if (cp >= 0x0300 && cp <= 0x036f) return true;
  // Combining Cyrillic
  if (cp >= 0x0483 && cp <= 0x0489) return true;
  // Hebrew combining
  if (cp >= 0x0591 && cp <= 0x05bd) return true;
  // Arabic combining
  if (cp >= 0x0610 && cp <= 0x061a) return true;
  if (cp >= 0x064b && cp <= 0x065f) return true;
  if (cp === 0x0670) return true;
  if (cp >= 0x06d6 && cp <= 0x06dc) return true;
  if (cp >= 0x06df && cp <= 0x06e4) return true;
  if (cp >= 0x06e7 && cp <= 0x06e8) return true;
  if (cp >= 0x06ea && cp <= 0x06ed) return true;
  // Syriac combining
  if (cp === 0x0711) return true;
  if (cp >= 0x0730 && cp <= 0x074a) return true;
  // Zero-width space, ZWNJ, ZWJ, direction marks
  if (cp >= 0x200b && cp <= 0x200f) return true;
  // Line/paragraph separators, direction overrides
  if (cp >= 0x2028 && cp <= 0x202e) return true;
  // Word joiner, invisible operators
  if (cp >= 0x2060 && cp <= 0x2064) return true;
  // Combining Marks for Symbols
  if (cp >= 0x20d0 && cp <= 0x20ff) return true;
  // Variation Selectors
  if (cp >= 0xfe00 && cp <= 0xfe0f) return true;
  // Combining Half Marks
  if (cp >= 0xfe20 && cp <= 0xfe2f) return true;
  // BOM / zero-width no-break space
  if (cp === 0xfeff) return true;
  // Variation Selectors Supplement
  if (cp >= 0xe0100 && cp <= 0xe01ef) return true;
  // Tag characters (used in flag subdivision sequences like 🏴󠁧󠁢󠁥󠁮󠁧󠁿)
  if (cp >= 0xe0020 && cp <= 0xe007f) return true;

  return false;
}

function isWide(cp: number): boolean {
  // Hangul Jamo
  if (cp >= 0x1100 && cp <= 0x115f) return true;
  // CJK Radicals, Kangxi, CJK Symbols
  if (cp >= 0x2e80 && cp <= 0x303e) return true;
  // Hiragana, Katakana, Bopomofo, Kanbun, CJK Compat
  if (cp >= 0x3040 && cp <= 0x33bf) return true;
  // CJK Extension A
  if (cp >= 0x3400 && cp <= 0x4dbf) return true;
  // CJK Unified Ideographs
  if (cp >= 0x4e00 && cp <= 0x9fff) return true;
  // Yi
  if (cp >= 0xa000 && cp <= 0xa4cf) return true;
  // Hangul Syllables
  if (cp >= 0xac00 && cp <= 0xd7af) return true;
  // CJK Compat Ideographs
  if (cp >= 0xf900 && cp <= 0xfaff) return true;
  // CJK Compat Forms, Small Forms
  if (cp >= 0xfe10 && cp <= 0xfe6f) return true;
  // Fullwidth Forms
  if (cp >= 0xff01 && cp <= 0xff60) return true;
  // Fullwidth Signs
  if (cp >= 0xffe0 && cp <= 0xffe6) return true;
  // Emoji with default emoji presentation (always width 2)
  if (inRanges(cp, EMOJI_PRESENTATION_RANGES)) return true;
  // CJK Extension B and beyond
  if (cp >= 0x20000 && cp <= 0x2ffff) return true;
  // CJK Extension G+
  if (cp >= 0x30000 && cp <= 0x3ffff) return true;

  return false;
}

/**
 * Returns true if a codepoint is a skin tone modifier (Fitzpatrick scale).
 * U+1F3FB through U+1F3FF.
 */
export function isSkinToneModifier(cp: number): boolean {
  return cp >= 0x1f3fb && cp <= 0x1f3ff;
}

/**
 * Returns true if a codepoint is a Regional Indicator Symbol.
 * U+1F1E6 through U+1F1FF. Pairs of these form flag emoji.
 */
export function isRegionalIndicator(cp: number): boolean {
  return cp >= 0x1f1e6 && cp <= 0x1f1ff;
}

/**
 * Return the total display width of a string in terminal columns.
 * Iterates by code point (handles surrogate pairs correctly).
 * Accounts for VS16 (U+FE0F) upgrading text-presentation emoji to width 2.
 * Accounts for grapheme clusters: skin tone modifiers, ZWJ sequences, and
 * regional indicator pairs are rendered as single glyphs by modern terminals.
 */
export function stringDisplayWidth(str: string): number {
  let width = 0;
  let prevCp = -1;
  let prevWasZWJ = false;
  let prevWasRI = false;
  for (const ch of str) {
    const cp = ch.codePointAt(0)!;
    if (cp === 0xfe0f && prevCp >= 0 && isTextPresentationEmoji(prevCp) && charDisplayWidth(prevCp) === 1) {
      // VS16 after a text-presentation emoji that is width 1: upgrade to width 2
      width += 1;
      prevCp = cp;
      prevWasZWJ = false;
      prevWasRI = false;
      continue;
    }
    const w = charDisplayWidth(cp);
    // Skin tone modifier after a wide character: part of same grapheme cluster
    if (isSkinToneModifier(cp) && prevCp >= 0 && charDisplayWidth(prevCp) === 2) {
      prevCp = cp;
      prevWasZWJ = false;
      prevWasRI = false;
      continue;
    }
    // Second regional indicator completes a flag pair: zero width
    if (isRegionalIndicator(cp) && prevWasRI) {
      prevCp = cp;
      prevWasZWJ = false;
      prevWasRI = false; // toggle off so third RI starts a new pair
      continue;
    }
    // Emoji after ZWJ: part of same grapheme cluster
    if (prevWasZWJ && w === 2) {
      prevCp = cp;
      prevWasZWJ = false;
      prevWasRI = false;
      continue;
    }
    width += w;
    prevWasZWJ = cp === 0x200d;
    prevWasRI = isRegionalIndicator(cp);
    prevCp = cp;
  }
  return width;
}

/**
 * Slice text from the start to fit within maxCols display columns.
 * Returns a substring whose display width is <= maxCols.
 * Never splits a surrogate pair, a wide character, or a grapheme cluster.
 */
export function sliceToWidth(text: string, maxCols: number): string {
  let cols = 0;
  let strIdx = 0;
  let prevCp = -1;
  let prevWasZWJ = false;
  let prevWasRI = false;
  // Track the string index of the start of the current grapheme cluster,
  // so we can back up if the cluster overflows.
  let clusterStartIdx = 0;
  let colsAtClusterStart = 0;
  for (const ch of text) {
    const cp = ch.codePointAt(0)!;
    let w: number;
    let partOfCluster = false;

    if (cp === 0xfe0f && prevCp >= 0 && isTextPresentationEmoji(prevCp) && charDisplayWidth(prevCp) === 1) {
      w = 1;
      partOfCluster = true;
    } else if (isSkinToneModifier(cp) && prevCp >= 0 && charDisplayWidth(prevCp) === 2) {
      w = 0;
      partOfCluster = true;
    } else if (isRegionalIndicator(cp) && prevWasRI) {
      w = 0;
      partOfCluster = true;
    } else if (prevWasZWJ && charDisplayWidth(cp) === 2) {
      w = 0;
      partOfCluster = true;
    } else {
      w = charDisplayWidth(cp);
    }

    if (cols + w > maxCols) {
      if (partOfCluster) {
        // Cluster overflows — back up to exclude the entire cluster
        strIdx = clusterStartIdx;
      }
      break;
    }

    if (!partOfCluster) {
      clusterStartIdx = strIdx;
      colsAtClusterStart = cols;
    }

    cols += w;
    strIdx += ch.length;
    prevWasZWJ = cp === 0x200d;
    prevWasRI = isRegionalIndicator(cp) && !partOfCluster;
    prevCp = cp;
  }
  return text.slice(0, strIdx);
}

/**
 * Slice text from the end to fit within maxCols display columns.
 * Returns a substring whose display width is <= maxCols.
 * Never splits a surrogate pair, a wide character, or a grapheme cluster.
 *
 * Uses grapheme cluster segmentation to avoid the complexity of
 * reverse-iterating with lookahead for ZWJ/RI/skin tone sequences.
 */
export function sliceFromEndToWidth(text: string, maxCols: number): string {
  // Segment into grapheme clusters by forward-iterating
  const clusters: string[] = [];
  let current = '';
  let prevCp = -1;
  let prevWasZWJ = false;
  let prevWasRI = false;

  for (const ch of text) {
    const cp = ch.codePointAt(0)!;
    let partOfCluster = false;

    if (cp === 0xfe0f && prevCp >= 0 && isTextPresentationEmoji(prevCp) && charDisplayWidth(prevCp) === 1) {
      partOfCluster = true;
    } else if (isSkinToneModifier(cp) && prevCp >= 0 && charDisplayWidth(prevCp) === 2) {
      partOfCluster = true;
    } else if (isRegionalIndicator(cp) && prevWasRI) {
      partOfCluster = true;
    } else if (prevWasZWJ && charDisplayWidth(cp) === 2) {
      partOfCluster = true;
    } else if (isZeroWidth(cp)) {
      partOfCluster = true;
    }

    if (partOfCluster) {
      current += ch;
    } else {
      if (current) clusters.push(current);
      current = ch;
    }

    prevWasZWJ = cp === 0x200d;
    prevWasRI = isRegionalIndicator(cp) && !partOfCluster;
    prevCp = cp;
  }
  if (current) clusters.push(current);

  // Iterate clusters from end
  let cols = 0;
  let count = 0;
  for (let i = clusters.length - 1; i >= 0; i--) {
    const w = stringDisplayWidth(clusters[i]!);
    if (cols + w > maxCols) break;
    cols += w;
    count++;
  }
  return clusters.slice(clusters.length - count).join('');
}

File: python/cellstate/width.py
"""
Unicode display width: how many terminal columns a character occupies (0, 1, or 2).
Width tables derived from Unicode 17.0 data.
"""

from __future__ import annotations
import bisect
from .emoji_data import EMOJI_PRESENTATION_RANGES, TEXT_PRESENTATION_EMOJI_RANGES

# Pre-compute sorted start/end lists for fast bisect lookups.
_EP_STARTS = [r[0] for r in EMOJI_PRESENTATION_RANGES]
_EP_ENDS   = [r[1] for r in EMOJI_PRESENTATION_RANGES]
_TP_STARTS = [r[0] for r in TEXT_PRESENTATION_EMOJI_RANGES]
_TP_ENDS   = [r[1] for r in TEXT_PRESENTATION_EMOJI_RANGES]


def _in_ranges(cp: int, starts: list[int], ends: list[int]) -> bool:
    """Binary search through sorted [start, end] inclusive ranges."""
    idx = bisect.bisect_right(starts, cp) - 1
    if idx < 0:
        return False
    return cp <= ends[idx]


def char_display_width(code_point: int) -> int:
    """
    Return the display width of a single Unicode code point.
    Returns 2 for fullwidth/wide characters (CJK, emoji, etc).
    Returns 0 for zero-width characters (combining marks, ZWJ, etc).
    Returns 1 for everything else.
    """
    if _is_zero_width(code_point):
        return 0
    # Terminal-specific overrides: 🈂 (U+1F202) and 🈷 (U+1F237) render as
    # width 2 in most terminals despite Emoji_Presentation=No.
    if code_point == 0x1F202 or code_point == 0x1F237:
        return 2
    if _is_wide(code_point):
        return 2
    return 1


def is_text_presentation_emoji(code_point: int) -> bool:
    """
    Returns True if the code point is a text-presentation emoji
    (Emoji=Yes but Emoji_Presentation=No). These are width 1 by default
    but become width 2 when followed by U+FE0F (VS16).
    """
    return _in_ranges(code_point, _TP_STARTS, _TP_ENDS)


def _is_zero_width(cp: int) -> bool:
    if 0x0300 <= cp <= 0x036F: return True   # Combining Diacritical Marks
    if 0x0483 <= cp <= 0x0489: return True   # Combining Cyrillic
    if 0x0591 <= cp <= 0x05BD: return True   # Hebrew combining
    if 0x0610 <= cp <= 0x061A: return True   # Arabic combining
    if 0x064B <= cp <= 0x065F: return True
    if cp == 0x0670: return True
    if 0x06D6 <= cp <= 0x06DC: return True
    if 0x06DF <= cp <= 0x06E4: return True
    if 0x06E7 <= cp <= 0x06E8: return True
    if 0x06EA <= cp <= 0x06ED: return True
    if cp == 0x0711: return True             # Syriac combining
    if 0x0730 <= cp <= 0x074A: return True
    if 0x200B <= cp <= 0x200F: return True   # Zero-width space, ZWNJ, ZWJ, direction marks
    if 0x2028 <= cp <= 0x202E: return True   # Line/paragraph separators, direction overrides
    if 0x2060 <= cp <= 0x2064: return True   # Word joiner, invisible operators
    if 0x20D0 <= cp <= 0x20FF: return True   # Combining Marks for Symbols
    if 0xFE00 <= cp <= 0xFE0F: return True   # Variation Selectors
    if 0xFE20 <= cp <= 0xFE2F: return True   # Combining Half Marks
    if cp == 0xFEFF: return True             # BOM / zero-width no-break space
    if 0xE0100 <= cp <= 0xE01EF: return True # Variation Selectors Supplement
    if 0xE0020 <= cp <= 0xE007F: return True # Tag characters
    return False


def _is_wide(cp: int) -> bool:
    if 0x1100 <= cp <= 0x115F: return True   # Hangul Jamo
    if 0x2E80 <= cp <= 0x303E: return True   # CJK Radicals, Kangxi, CJK Symbols
    if 0x3040 <= cp <= 0x33BF: return True   # Hiragana, Katakana, Bopomofo, etc.
    if 0x3400 <= cp <= 0x4DBF: return True   # CJK Extension A
    if 0x4E00 <= cp <= 0x9FFF: return True   # CJK Unified Ideographs
    if 0xA000 <= cp <= 0xA4CF: return True   # Yi
    if 0xAC00 <= cp <= 0xD7AF: return True   # Hangul Syllables
    if 0xF900 <= cp <= 0xFAFF: return True   # CJK Compat Ideographs
    if 0xFE10 <= cp <= 0xFE6F: return True   # CJK Compat Forms, Small Forms
    if 0xFF01 <= cp <= 0xFF60: return True   # Fullwidth Forms
    if 0xFFE0 <= cp <= 0xFFE6: return True   # Fullwidth Signs
    if _in_ranges(cp, _EP_STARTS, _EP_ENDS): return True  # Emoji with default emoji presentation
    if 0x20000 <= cp <= 0x2FFFF: return True # CJK Extension B and beyond
    if 0x30000 <= cp <= 0x3FFFF: return True # CJK Extension G+
    return False


def is_skin_tone_modifier(cp: int) -> bool:
    """Returns True if a codepoint is a skin tone modifier (Fitzpatrick scale, U+1F3FB-U+1F3FF)."""
    return 0x1F3FB <= cp <= 0x1F3FF


def is_regional_indicator(cp: int) -> bool:
    """Returns True if a codepoint is a Regional Indicator Symbol (U+1F1E6-U+1F1FF)."""
    return 0x1F1E6 <= cp <= 0x1F1FF


def string_display_width(s: str) -> int:
    """
    Return the total display width of a string in terminal columns.
    Iterates by code point; accounts for VS16, skin tone modifiers,
    ZWJ sequences, and regional indicator pairs.
    """
    width = 0
    prev_cp = -1
    prev_was_zwj = False
    prev_was_ri = False
    for ch in s:
        cp = ord(ch)
        if (cp == 0xFE0F and prev_cp >= 0
                and is_text_presentation_emoji(prev_cp)
                and char_display_width(prev_cp) == 1):
            # VS16 after a text-presentation emoji: upgrade to width 2
            width += 1
            prev_cp = cp
            prev_was_zwj = False
            prev_was_ri = False
            continue
        w = char_display_width(cp)
        if is_skin_tone_modifier(cp) and prev_cp >= 0 and char_display_width(prev_cp) == 2:
            prev_cp = cp
            prev_was_zwj = False
            prev_was_ri = False
            continue
        if is_regional_indicator(cp) and prev_was_ri:
            prev_cp = cp
            prev_was_zwj = False
            prev_was_ri = False
            continue
        if prev_was_zwj and w == 2:
            prev_cp = cp
            prev_was_zwj = False
            prev_was_ri = False
            continue
        width += w
        prev_was_zwj = (cp == 0x200D)
        prev_was_ri = is_regional_indicator(cp)
        prev_cp = cp
    return width


def slice_to_width(text: str, max_cols: int) -> str:
    """
    Slice text from the start to fit within max_cols display columns.
    Never splits a surrogate pair, a wide character, or a grapheme cluster.
    """
    cols = 0
    str_idx = 0
    prev_cp = -1
    prev_was_zwj = False
    prev_was_ri = False
    cluster_start_idx = 0

    for ch in text:
        cp = ord(ch)
        part_of_cluster = False
        w: int

        if (cp == 0xFE0F and prev_cp >= 0
                and is_text_presentation_emoji(prev_cp)
                and char_display_width(prev_cp) == 1):
            w = 1
            part_of_cluster = True
        elif is_skin_tone_modifier(cp) and prev_cp >= 0 and char_display_width(prev_cp) == 2:
            w = 0
            part_of_cluster = True
        elif is_regional_indicator(cp) and prev_was_ri:
            w = 0
            part_of_cluster = True
        elif prev_was_zwj and char_display_width(cp) == 2:
            w = 0
            part_of_cluster = True
        else:
            w = char_display_width(cp)

        if cols + w > max_cols:
            if part_of_cluster:
                str_idx = cluster_start_idx
            break

        if not part_of_cluster:
            cluster_start_idx = str_idx

        cols += w
        str_idx += len(ch)
        prev_was_zwj = (cp == 0x200D)
        prev_was_ri = is_regional_indicator(cp) and not part_of_cluster
        prev_cp = cp

    return text[:str_idx]


def slice_from_end_to_width(text: str, max_cols: int) -> str:
    """
    Slice text from the end to fit within max_cols display columns.
    Never splits a surrogate pair, a wide character, or a grapheme cluster.
    """
    # Segment into grapheme clusters by forward-iterating
    clusters: list[str] = []
    current = ""
    prev_cp = -1
    prev_was_zwj = False
    prev_was_ri = False

    for ch in text:
        cp = ord(ch)
        part_of_cluster = False

        if (cp == 0xFE0F and prev_cp >= 0
                and is_text_presentation_emoji(prev_cp)
                and char_display_width(prev_cp) == 1):
            part_of_cluster = True
        elif is_skin_tone_modifier(cp) and prev_cp >= 0 and char_display_width(prev_cp) == 2:
            part_of_cluster = True
        elif is_regional_indicator(cp) and prev_was_ri:
            part_of_cluster = True
        elif prev_was_zwj and char_display_width(cp) == 2:
            part_of_cluster = True
        elif _is_zero_width(cp):
            part_of_cluster = True

        if part_of_cluster:
            current += ch
        else:
            if current:
                clusters.append(current)
            current = ch

        prev_was_zwj = (cp == 0x200D)
        prev_was_ri = is_regional_indicator(cp) and not part_of_cluster
        prev_cp = cp

    if current:
        clusters.append(current)

    # Iterate clusters from end
    cols = 0
    count = 0
    for cluster in reversed(clusters):
        w = string_display_width(cluster)
        if cols + w > max_cols:
            break
        cols += w
        count += 1

    return "".join(clusters[len(clusters) - count:])
